### PR TITLE
feat: source-block fence shortcut and multi-block code toggle in inline editor

### PR DESCRIPTION
**Pull request by sh1nj1**
## Summary

Fixes two source-block (code-block) shortcomings in the Creative inline Lexical editor.

**1. Typing a \`\`\` fence now opens a source block.**
Previously the built-in `@lexical/markdown` `CODE` transformer only fired on `` ``` `` **plus a trailing space**. Users type the fence and press **Enter**, which left the line as plain text. A new `registerCodeFenceShortcut` transform converts a paragraph whose whole text is `` ``` `` (optionally `` ```lang ``) into a code block as it's typed, dropping the caret inside — the standard rich-editor behavior. Prose that merely contains backticks (e.g. `` run ```code``` inline ``) is left untouched.

**2. The toolbar code-block button now applies to every selected line.**
`toggleCodeBlock` only processed the anchor's top-level block, so selecting several paragraphs and hitting the button converted **only the last line**. It now merges every selected top-level block into a single code block (one line per block). Toggling a code selection off expands each block back to one paragraph per line and restores a valid caret. Mixed selections (headings/quotes/lists/paragraphs) are absorbed as plain text so the button always has a predictable effect on a range.

## Changes
- `lib/lexical/code_fence_shortcut.js` — new: bare-fence → code-block transform (guarded to the actively-edited block so bulk import/paste isn't rewritten).
- `lib/lexical/code_block_toggle.js` — new: `$toggleCodeBlockForSelection()`, multi-block toggle logic extracted from the component.
- `components/InlineLexicalEditor.jsx` — `toggleCodeBlock` delegates to the new module.
- `components/plugins/markdown_shortcuts_plugin.jsx` — registers the fence shortcut alongside the existing markdown shortcuts.

## Testing
- New headless Lexical jest tests for both modules (8 cases).
- Full JS suite: **487 passed / 45 suites**.
- `npm run build` (esbuild) clean.

JS-only change; Ruby pre-push (`rake test`/rubocop) skipped as irrelevant.

### CODE REVIEW

File: engines/collavre/app/javascript/lib/lexical/code_fence_shortcut.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    // Only convert the block the user is actively typing in — a collapsed caret
+    // inside this paragraph. This keeps bulk operations (import, paste, programmatic
+    // edits) that happen to produce a ``` paragraph elsewhere from being rewritten.
+    const selection = $getSelection()
+    if (!$isRangeSelection(selection) || !selection.isCollapsed()) return
+    if (selection.anchor.getNode().getTopLevelElement() !== node) return
+
+    const language = match[1] || undefined
+    const codeNode = $createCodeNode(language)
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_fence_shortcut.js]: Fixed in 85ab9d30. When the fence carries a language, the new CodeNode is now recorded via `markLanguageResolved(editor, codeNode.getKey())`, mirroring the import path (InlineLexicalEditor.jsx:170). The detect transform (`isLanguageResolved` guard) now honors an explicit ```javascript verbatim instead of re-detecting and relabeling it. Regression tests added: `marks an explicit fence language resolved…` and `does not mark a bare ``` fence resolved`.

File: engines/collavre/app/javascript/lib/lexical/code_fence_shortcut.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * The built-in @lexical/markdown CODE transformer only fires on ``` + a trailing
+ * space, so pressing Enter after the fence (what users actually do) left the
+ * paragraph as plain text. This transform reacts to the fence itself.
+ *
+ * Returns the editor.registerNodeTransform teardown so callers can clean up.
+ */
+export function registerCodeFenceShortcut(editor) {
+  return editor.registerNodeTransform(ParagraphNode, (node) => {
+    const match = node.getTextContent().match(FENCE_REGEX)
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_fence_shortcut.js]: Intended tradeoff, not fixing. The requirement is immediate conversion — "typing a leading fence auto-converts to a source block" — which is the standard rich-editor behavior (Notion/Slack/Discord all convert the moment ``` is complete). Inline language typing (```ruby as you type) is fundamentally incompatible with immediate conversion: a per-character node transform has no signal that more of a language token is still coming. So during live typing a bare ``` converts and the caret enters the empty block; the block language is then chosen from the block UI. match[1] is still meaningful for the paste/programmatic path where the full ```lang lands at once — now also correctly marked resolved by 85ab9d30. Clarified this in the FENCE_REGEX comment.

File: engines/collavre/app/javascript/lib/lexical/code_block_toggle.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    // end of the expanded text so Lexical keeps a valid selection.
+    if (lastParagraph) lastParagraph.selectEnd()
+    return
+  }
+
+  // Otherwise → merge every selected block's text into a single code block.
+  const content = topLevels.map((node) => node.getTextContent()).join("\n")
+  const codeNode = $createCodeNode()
+  codeNode.append($createTextNode(content))
+  topLevels[0].replace(codeNode)
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_block_toggle.js]: Fixed in 63cc022e. Confirmed the regression: the old toolbar path returned for any non-paragraph top level, so a table cell caret (which `getTopLevelElement()` resolves to the root `TableNode`) was left alone. The new merge path dropped that guard to absorb headings/quotes/lists as text, which unintentionally made a table-cell selection flatten the whole table.

The same destructive-removal class also hit top-level `DecoratorNode`s registered in this editor — `ImageNode`, `VideoNode`, `AttachmentNode` (`trailing_paragraph.js:14` treats them as top-level media): their `getTextContent()` is empty and the merge would `.remove()` them.

Fix: the merge now excludes structural blocks — `!$isTableNode(node) && !$isDecoratorNode(node)` — so only real text blocks (paragraph/heading/quote/list/code) fold into the code block; tables and media are left in place. If a selection contains only non-mergeable blocks the toggle is a no-op (matches the old paragraph-only guard). Regression tests added: a caret inside a table cell keeps the table intact, and a media block survives a surrounding-text merge.

File: engines/collavre/app/javascript/lib/lexical/code_block_toggle.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    mergeable[index].remove()
+  }
+  codeNode.selectEnd()
+}
+
+// A block can be folded into a code block only if its text content faithfully
+// represents it. Tables (tabular structure) and decorator media (no text) do
+// not, so they are excluded from the merge to avoid destroying content.
+function $isMergeableTextBlock(node) {
+  return !$isTableNode(node) && !$isDecoratorNode(node)
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_block_toggle.js]: Fixed in f45ba8db. Confirmed via Lexical's own source (`Lexical.dev.mjs:9549` documents that `getTopLevelElement()` inside a TableCellNode returns the cell's first child, not the root TableNode), so cell content slipped past the `$isTableNode(node)` check. The merge now also excludes any block with a TableNode ancestor — `node.getParents().some($isTableNode)` — so table-scoped blocks are never merged with document-root blocks. Regression test added: a range walking through a table interior leaves the cell text in place and never folds it into a code block.

File: engines/collavre/app/javascript/lib/lexical/code_block_toggle.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  // their text content is empty or a flattening of tabular data, so absorbing
+  // them would silently destroy the table/media. Leave those blocks in place
+  // and only merge the real text blocks (matches the old paragraph-only guard,
+  // widened to headings/quotes/lists but not to structural containers).
+  const mergeable = topLevels.filter($isMergeableTextBlock)
+  if (mergeable.length === 0) return
+  const content = mergeable.map((node) => node.getTextContent()).join("\n")
+  const codeNode = $createCodeNode()
+  codeNode.append($createTextNode(content))
+  mergeable[0].replace(codeNode)
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_block_toggle.js]: Fixed in a6d41e1b, taking the second option (merge only contiguous text blocks). A skipped structural block now splits the selected text into runs, and each contiguous run of text blocks becomes its own code block while the structural block stays in place — later content is never pulled across it. So `paragraph → image/table → paragraph` yields `code → image/table → code`, preserving document order. The media and table regression tests now assert the `[code, structural, code]` ordering and per-run content.

File: engines/collavre/app/javascript/lib/lexical/code_block_toggle.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+// Distinct top-level blocks the selection touches, in document order. getNodes()
+// returns the range's nodes in document order; each maps to its owning top-level
+// element. Falls back to the anchor's block for a collapsed selection.
+function $selectedTopLevelBlocks(selection) {
+  const seen = new Set()
+  const blocks = []
+  selection.getNodes().forEach((node) => {
+    const topLevel = node.getTopLevelElement()
```
[the remark this answers is not in the document]
Comment (sh1nj1) [reply engines/collavre/app/javascript/lib/lexical/code_block_toggle.js:102]: Fixed in 1f641a70. Confirmed the regression: `ListNode` is the top-level element for every bullet (this editor treats `topLevel` as the ListNode at `InlineLexicalEditor.jsx:647`), so a selection inside one bullet top-levels to the whole list; `flushRun()` then reads `getTextContent()` across all bullets and replaces the entire `ListNode`, absorbing unselected siblings.

`ListItemNode` is not a shadow root (only `TableCellNode` is registered as one here), so list content always resolves to the outermost `ListNode` and a single `$isListNode(node)` guard also covers nested lists — no parent walk needed.

Fix: lists are now treated as structural (like tables/media) and left in place — `$isMergeableTextBlock` returns false for a `ListNode`, so a list splits the merge run instead of being folded. Item-granular folding (split the list, convert only the selected bullets, keep the rest) is intentionally out of scope here — every prior broadening of this merge produced a new data-loss edge case, and lists were not part of the original requirement; a list is converted to code via a ``` fence instead. Regression tests added: one bullet selected leaves the list whole, and a list inside a range yields `[code, list, code]` with document order preserved.
